nexus/src/app/sagas/instance_update/mod.rs

```diff
@@ -1189,9 +1189,37 @@ async fn siu_commit_instance_updates(
     // try to start a new "actual" instance update saga that inherits our lock.
     // This way, we could also avoid computing updates required twice.
     // But, I'm a bit sketched out by the implications of not committing update
-    // and dropping the lock in the same operation. This deserves more thought...
-    if let Err(error) =
-        chain_update_saga(&sagactx, authz_instance, serialized_authn).await
+    // and dropping the lock in the same operation. This deserves more
+    // thought...
+
+    // Fetch the state from the database again to see if we should immediately
+    // run a new saga.
+    let new_state = match osagactx
+        .datastore()
+        .instance_fetch_all(&opctx, &authz_instance)
+        .await
+    {
+        Ok(s) => s,
+        // Do NOT unwind the rest of the saga if this fails, since we've
+        // already committed the update.
+        Err(e) => {
+            warn!(log,
+                "instance update: failed to fetch latest state after update";
+                "instance_id" => %instance_id,
+                "error" => %e,
+            );
+            nexus.background_tasks.task_instance_updater.activate();
+            return Ok(());
+        }
+    };
+
+    if let Err(error) = chain_update_saga(
+        &sagactx,
+        authz_instance,
+        serialized_authn,
+        &new_state,
+    )
+    .await
```

Comment:
It looks like this was here before and I don't know that this is wrong but it makes me nervous when we do more than one thing in a saga action because it makes it harder to test that the action is idempotent (i.e., it means our automated tests generally won't catch non-idempotent actions because they never _half_-do an action).  So I guess the question is: is this definitely idempotent?  And would it be reasonable to separate it anyway so that the tests make sure of it?

Reply:
Opened #6630 to address this, thanks for the suggestion!